### ISSUE TITLE
CMS being instanced twice on current version

### ISSUE THREAD

**Issue by bobmitch** [HoltBosse/Alba#193]
Some of the dumpster fire DB stuff done in https://github.com/HoltBosse/Alba/commit/b0675e167f27e2276e49795c1330a10a4f6fea85 actually causes a circular reference to the CMS singleton instantiation via the DB:: calls. 

Every query in Configuration and CMS core constructor **needs** to be done via Instance() calls or $this in order to not have CMS create a second version of itself.

Proof via print_r ('boo') in core CMS.php:

```php
public final static function Instance(){
		
		if (self::$instance === null) {
			print_r ('boo'); echo "<hr>";
			self::$instance = new CMS();
		}
		return self::$instance;
	}
```

results in:

<img width="391" alt="image" src="https://user-images.githubusercontent.com/23583515/189267819-0bfa0c04-8849-42f8-bf7f-7c8420fef0da.png">

**Comment by bobmitch** [comment]
User class was noted to cause this issue waaaaay back in the early CMS days, but I should have realized that DB would cause the same problem while reviewing @workaj s cleanup.  Will hit in the morning and make package for updating wherever needs it.

This shouldn't break _too_ much, but can cause weird issues with messages and session stuff that makes debugging _nightmarish_.

**Comment by bobmitch** [comment]
TLDR: Cant use autoloaded classes that refer back to the CMS object within it's own constructor or it will make multiple copies of itself in spite of the singleton's code trying to prevent such a thing.

**Comment by workaj** [comment]
fun stuff. perhaps stuff that says needs testing should be tested first.... 🤔. this might be the source of some interesting message situations in that project i work a lot on

**Comment by workaj** [comment]
alternatively something like https://github.com/HoltBosse/Alba/blob/master/core/cms.php#L217 could be done if we actually wanted to use DB in cms.php, but is probably way more trouble and issues than it is worth

**Comment by bobmitch** [comment]
> alternatively something like https://github.com/HoltBosse/Alba/blob/master/core/cms.php#L217 could be done if we actually wanted to use DB in cms.php, but is probably way more trouble and issues than it is worth

It's not even the _whole_ of cms.php - just the construction function. Think the Class method is fine if you can live with raw PHP PDO stuff in the constructor and classes it autoloads to do it's magic. ;) 

Will start to document some tests that should be done for each release. Testing for generating twin CMS instances is top of the list for anything that touches cms.php. Another downside of singleton patterns is their aversion to being easily unit tested (https://stackoverflow.com/questions/2085953/unit-testing-with-singletons).

**Comment by workaj** [comment]
> SNIP
> Will start to document some tests that should be done for each release. Testing for generating twin CMS instances is top of the list for anything that touches cms.php. Another downside of singleton patterns is their aversion to being easily unit tested (https://stackoverflow.com/questions/2085953/unit-testing-with-singletons).

github action the loads the cms and runs tests would probably be handy

**Comment by bobmitch** [comment]
Note for future self:  https://testing.googleblog.com/2008/08/by-miko-hevery-so-you-join-new-project.html?showComment=1395380128737#c1173766245909651201

Instead of using DB:: in core classes, expose a DB Instance function similar to CMS for 'internal' use, so DB calls would become:

```php
$foo = DB::Instance()->fetch('select * from bar');
```

Theoretically this would, at the expense of more succinct code, allow unit testing of individual pieces of code since they explicitly define their _collaborators_ instead of assuming they magically exist.

Food for thought.

**Comment by workaj** [comment]
fixed https://github.com/HoltBosse/Alba/commit/cd79b90d93c1b426d8b177c71f9bb835686b6099
